src/xkeys.rs

```diff
@@ -0,0 +1,371 @@
+use std::fmt::{self, Debug};
+
+const XKEY_VERSION_V1: &[u8] = b"xkv1";
+
+use crypto_box::{PublicKey, SecretKey};
+use rand::{CryptoRng, Rng, RngCore};
+
+/// The main interface used for reading and writing _nkey-encoded_ curve key
+/// pairs. Instances of this type cannot be cloned.
+pub struct XKey {
+    public: PublicKey,
+    secret: Option<SecretKey>,
+}
+
+impl Debug for XKey {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "XKey")
+    }
+}
+
+impl XKey {
+    /// Creates a new xkey.
+    ///
+    /// NOTE: This is not available if using on a wasm32-unknown-unknown target due to the lack of
+    /// rand support. Use [`new_from_raw`](XKey::new_from_raw) instead
+    #[cfg(not(target_arch = "wasm32"))]
+    pub fn new() -> Self {
+        Self::new_with_rand(&mut rand::rngs::OsRng)
+    }
+
+    /// Create a new xkey pair from a random generator
+    ///
+    /// NOTE: These generator should be a cryptographically secure random source.
+    ///
+    /// NOTE: This is not available if using on a wasm32-unknown-unknown target due to the lack of
+    /// rand support. Use [`new_from_raw`](XKey::new_from_raw) instead
+    #[cfg(not(target_arch = "wasm32"))]
+    pub fn new_with_rand(rand: &mut (impl CryptoRng + RngCore)) -> Self {
+        Self::new_from_raw(rand.gen())
+    }
+
+    /// Create a new xkey pair using a pre-existing set of random bytes.
+    ///
+    /// NOTE: These bytes should be generated from a cryptographically secure random source.
+    pub fn new_from_raw(random_bytes: [u8; 32]) -> Self {
+        let private = SecretKey::from_bytes(random_bytes);
+        Self {
+            public: private.public_key(),
+            secret: Some(private),
+        }
+    }
+
+    /// Attempts to produce a public-only xkey from the given encoded public key string
+    pub fn from_public_key(source: &str) -> Result<Self> {
+        let source_bytes = source.as_bytes();
+        let raw = decode_raw(source_bytes)?;
+
+        let (prefix, rest) = raw.split_first().ok_or(err!(VerifyError, "Empty key"))?;
+        if *prefix != PREFIX_BYTE_CURVE {
+            Err(err!(
+                InvalidPrefix,
+                "Not a valid public key prefix: {}",
+                raw[0]
+            ))
+        } else {
+            let public = PublicKey::try_from(rest)
+                .map_err(|_| err!(VerifyError, "Could not read public key"))?;
+
+            Ok(Self {
+                public,
+                secret: None,
+            })
+        }
+    }
+
+    /// Attempts to produce a full xkey pair from the given encoded seed string
+    pub fn from_seed(source: &str) -> Result<Self> {
+        if source.len() != ENCODED_SEED_LENGTH {
+            let l = source.len();
+            return Err(err!(InvalidSeedLength, "Bad seed length: {}", l));
+        }
+
+        let source_bytes = source.as_bytes();
+        let raw = decode_raw(source_bytes)?;
+
+        let b1 = raw[0] & 248;
```

Comment:
A lot of this code is copied from the plain nkeys logic. It should probably be abstracted into a function that both sets of logic use

Reply:
Extracted to a `from_seed` function that returns the type and seed